### PR TITLE
Add hasInteractedByUser getter in FormField

### LINKED ISSUE #131538: Add getter for FormField's _hasInteractedByUser
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I'm extending the `FormFieldState` class to create a custom type of form. I've overridden the `save` method to call another method in the codebase, whenever `save` is triggered from the wrapping `Form`. I only want to call this other method if the field has been interacted with.

`FormField` has a private flag for this, `_hasInteractedByUser`, which could provide the required information. A way to currently solve this is to override `didChange` and `reset` to keep track of my own boolean flag.

### Proposal

I would like `FormFieldState` to expose a boolean getter for the value of `_hasInteractedByUser`. I see two cases in which this could be useful.

1. When accessing the `FormFieldState` via a key.

```dart
final key = GlobalKey<FormFieldState<String>>;

TextFormField(
  key: key,
),

final changed = key.currentState!.hasInteractedByUser;
```

2. When extending `FormFieldState`.

```dart
@override
void save() {
  super.save();
  if(hasInteractedByUser) {
    doSomethingAwesome();
  }
}
```

### PR DESCRIPTION
**Pull request by dumazy**
Adds a getter to access the value of the private `RestorableBool _hasInteractedByUser`.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes #131538

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/form.dart [hunk 1]
```diff
@@ -480,6 +480,11 @@ class FormFieldState<T> extends State<FormField<T>> with RestorationMixin {
   /// True if this field has any validation errors.
   bool get hasError => _errorText.value != null;
 
+  /// True if the user has modified the value of this field. This only updates
+  /// to true once [didChange] has been called and resets to false when [reset]
+  /// is called.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/form.dart:485]:
```suggestion
  /// Returns true if the user has modified the value of this field.
  ///
  /// This only updates to true once [didChange] has been called and resets to
  /// false when [reset] is called.
```

**Review by chunhtai — APPROVED** [review]
LGTM, except for the suggestion

File: packages/flutter/test/widgets/form_test.dart [hunk 2]
```diff
@@ -854,4 +854,84 @@ void main() {
...
+    );
+
+    await tester.pumpWidget(widget);
+
+    await tester.enterText(find.byType(TextFormField), 'foo');
+    await tester.pump();
+
+    fieldKey.currentState!.reset();
+
+    expect(fieldKey.currentState!.hasInteractedByUser, isFalse);
```
Comment (justinmc) [inline packages/flutter/test/widgets/form_test.dart:935]: Nit: Similarly maybe put an expectation before and after `enterText`.

**Review by justinmc — APPROVED** [review]
LGTM with nits.  I think it makes sense for exposing this based on your arguments in the issue.  Thanks for providing those example usages.

File: packages/flutter/test/widgets/form_test.dart (hunk 2 above)
Comment (dumazy) [reply packages/flutter/test/widgets/form_test.dart:941]: Updated the test to validate before and after entering text. Thanks for the suggestion :)

**Comment by dumazy** [comment]
Looks like the tests of widget_span timed out and this failing the test. How do we proceed from here?

**Comment by justinmc** [comment]
I'll rerun it and it should pass.

**Pull request merged**